src/Workspaces/Core/Portable/Extensions/ValueUsageInfo.cs

```diff
@@ -52,18 +53,38 @@ internal enum ValueUsageInfo
         /// Represents a value read or write.
         /// For example, passing an argument to a "ref" parameter.
         /// </summary>
-        ReadableWritableRef = ReadableRef | WritableRef
+        ReadableWritableReference = ReadableReference | WritableReference
     }
 
     internal static class ValueUsageInfoExtensions
     {
-        public static bool ContainsReadOrReadableRef(this ValueUsageInfo valueUsageInfo)
-            => (valueUsageInfo & (ValueUsageInfo.Read | ValueUsageInfo.ReadableRef)) != 0;
+        public static bool ContainsReadOrReadableReference(this ValueUsageInfo valueUsageInfo)
+            => (valueUsageInfo & (ValueUsageInfo.Read | ValueUsageInfo.ReadableReference)) != 0;
 
-        public static bool ContainsWriteOrWritableRef(this ValueUsageInfo valueUsageInfo)
-            => (valueUsageInfo & (ValueUsageInfo.Write | ValueUsageInfo.WritableRef)) != 0;
+        public static bool ContainsWriteOrWritableReference(this ValueUsageInfo valueUsageInfo)
+            => (valueUsageInfo & (ValueUsageInfo.Write | ValueUsageInfo.WritableReference)) != 0;
 
-        public static bool ContainsNonReadWriteRef(this ValueUsageInfo valueUsageInfo)
-            => (valueUsageInfo & ValueUsageInfo.NonReadWriteRef) != 0;
+        public static bool ContainsNonReadWriteReference(this ValueUsageInfo valueUsageInfo)
+            => (valueUsageInfo & ValueUsageInfo.NonReadWriteReference) != 0;
+
+        public static ImmutableArray<string> ToValues(this ValueUsageInfo valueUsageInfo)
+        {
+            if (valueUsageInfo == ValueUsageInfo.None)
+            {
+                return ImmutableArray<string>.Empty;
+            }
+
+            var builder = ImmutableArray.CreateBuilder<string>();
+            foreach (ValueUsageInfo value in Enum.GetValues(typeof(ValueUsageInfo)))
+            {
+                bool singleBitIsSet = (value & (value - 1)) == 0;
+                if (singleBitIsSet && (valueUsageInfo & value) != 0)
+                {
+                    builder.Add(value.ToString());
+                }
+            }
+
+            return builder.ToImmutable();
```

Comment:
it would probably be *very* good to cache this.  Also, is this passed back to the editor as an interface type (like IList/IReadOnlyList?).  If so, we should cache in that form so we don't box on converting the immutablearray to the interface.

Reply:
This is now cached in the FAR column which invokes this method.

---
In reply to: [220754589](https://github.com/dotnet/roslyn/pull/30155#discussion_r220754589) [](ancestors = 220754589)